operator/pkg/ciliumidentity/controller_test.go

```diff
@@ -214,6 +218,113 @@ func verifyCIDUsageInCES(ctx context.Context, fakeClient *k8sClient.FakeClientse
 	return nil
 }
 
+// TestNoIDAllocatorLeaks ensures that when pods and their CIDs are deleted,
+// the CID controller returns the used CIDs to the available pool.
+func TestNoIDAllocatorLeaks(t *testing.T) {
+	const (
+		uniqueIDsToCreate     = 8
+		firstCreatePods       = 5
+		deletePods            = 3
+		createNewPodsStartIdx = 5
+		createNewPodsEndIdx   = 7
+
+		waitTimeout = 20 * time.Second
+	)
+	var nsList []*slim_corev1.Namespace
+	for i := 0; i < uniqueIDsToCreate; i++ {
+		nsList = append(nsList, testCreateNSObj(fmt.Sprintf("ns-%d", i), nil))
+	}
+
+	var podList []*slim_corev1.Pod
+	for i := 0; i < uniqueIDsToCreate; i++ {
+		podList = append(podList, testCreatePodObj("pod1", fmt.Sprintf("ns-%d", i), testLbsA, nil))
+	}
+
+	var cidList []*capi_v2.CiliumIdentity
+	for i := 0; i < uniqueIDsToCreate; i++ {
+		cidList = append(cidList, testCreateCIDObjNs("1000", podList[i], nsList[i]))
+	}
+
+	// Use a custom ID range to verify if CIDs are leaked, without having to
+	// create 65k (default) allowed CIDs.
+	customIDRange := idRange{
+		MinIDValue: idpool.ID(1000),
+		MaxIDValue: idpool.ID(1005),
+	}
+	// Start test hive.
+	cidResource, _, fakeClient, _, h := initHiveTest(t, true, customIDRange)
+	ctx, cancelCtxFunc := context.WithCancel(context.Background())
+	tlog := hivetest.Logger(t)
+	if err := h.Start(tlog, ctx); err != nil {
+		t.Fatalf("starting hive encountered an error: %s", err)
+	}
+	defer func() {
+		if err := h.Stop(tlog, ctx); err != nil {
+			t.Fatalf("stopping hive encountered an error: %v", err)
+		}
+		cancelCtxFunc()
+	}()
+
+	for i := range uniqueIDsToCreate {
+		if _, err := fakeClient.Slim().CoreV1().Namespaces().Create(ctx, nsList[i], metav1.CreateOptions{}); err != nil {
+			t.Fatalf("create namespace: %v", err)
+		}
+	}
+
+	store, err := (*cidResource).Store(ctx)
+	if err != nil {
+		t.Fatalf("unexpected error while getting CID store: %s", err)
+	}
+
+	// Create the first 5 pods.
+	for i := range firstCreatePods {
+		if _, err := fakeClient.Slim().CoreV1().Pods(podList[i].Namespace).Create(ctx, podList[i], metav1.CreateOptions{}); err != nil {
+			t.Fatalf("create pod: %v", err)
```

Comment:
Be more descriptive in the error message; this signature matches below and may lead to incorrect conclusions if the line number is not considered.

nit: Also, logged messages such as these should be complete sentences

Reply:
Done